Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29443 Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s**
* #29747 Refactoring torch.distributed.rpc tests to have Mixins and Fixtures

https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_model_parallel` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: [D18283849](https://our.internmc.facebook.com/intern/diff/D18283849/)

xush6528 committed:
Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s

https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_model_parallel` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: [D18283849](https://our.internmc.facebook.com/intern/diff/D18283849/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s"


https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_model_parallel` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: [D18283849](https://our.internmc.facebook.com/intern/diff/D18283849/)

[ghstack-poisoned]

mrshenli commented on test/process_group_rpc_test.py:
```diff
@@ -0,0 +1,40 @@
+    return decorator
+
+
+class ProcessGroupRpcTest(rpc_test.RpcTest):
```
This should inherit from `ProcessGroupRpcAgentTestFixture`, as it uses `rpc_agent_options` property?

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -1,24 +1,37 @@
     """Registers a new RPC backend.
 
     Arguments:
         backend (str): backend string to identify the handler.
```
`backend_name` or rename the argument as `backend` to make them consistent

xush6528 replied:
Updating the comment to `backend_name`, calling it "name" because it's a string, not a backend enum type member.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -1,24 +1,37 @@
+        construct_rpc_agent_options_handler (function):
+            Handler that is invoked when
+            rpc_backen.construct_rpc_agent_options(**dict) is called.
+        Handler that is invoked when the
```
Remove this?

xush6528 replied:
Removed.

mrshenli commented on test/test_dist_autograd_fork.py:
```diff
@@ -9,9 +9,10 @@
 
 
-class DistAutogradTestWithFork(MultiProcessTestCase, DistAutogradTest):
+class DistAutogradTestWithFork(MultiProcessTestCase, ProcessGroupRpcAgentTestFixture, DistAutogradTest):
```
Question: if we add this fixture here, how are going to extend this test for other backend?

xush6528 replied:
Moving this `RpcAgentTestFixture` to `class DistAutogradTest(RpcAgentTestFixture)`.

mrshenli commented on torch/csrc/distributed/rpc/types.h:
```diff
@@ -51,6 +51,11 @@ struct TORCH_API SerializedPyObj final {
   const std::vector<at::Tensor> tensors_;
 };
 
+struct RpcAgentOptions {
```
Question: why this is here instead of in `RpcAgent.h`?

xush6528 replied:
I think it's better to put it under `RpcAgent.h`. Moved.

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -50,29 +51,51 @@ def init_model_parallel(
-            rpc_timeout (datetime.timedelta): Timeout for RPCs. Defaults to 10 seconds.
-                0 means infinity.
+            self_rank (int): a globally unique id/rank of this node.
+            world_size (int): The number of workers in the group.
```
adding one line for `rpc_agent_options`?

xush6528 replied:
Added.

xush6528 committed:
Update on "Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s"


https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_model_parallel` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: [D18283849](https://our.internmc.facebook.com/intern/diff/D18283849/)

[ghstack-poisoned]

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -1016,37 +1009,6 @@ def test_backward_autograd_engine_error(self):
-    @unittest.skipIf(TEST_CONFIG.rpc_backend_name == "PROCESS_GROUP",
-                     "Skipping this test temporarily since ProcessGroupAgent does not report errors on node failures")
-    @dist_init(clean_shutdown=False)
-    def test_backward_node_failure(self):
```
Why removing these two tests?

xush6528 replied:
The error message in these 2 tests are ThriftClient error messages. I moved these 2 into fb/ folder to make it invisible from OSS.

xush6528 replied:
We can definitely move it back to the shared test suite once we go through all failure cases and make sure in those cases, the common OSS layer always throws the same exception type and message for different backends.

mrshenli commented on test/test_dist_autograd_fork.py:
```diff
@@ -1,21 +1,23 @@
 
-class DistAutogradTestWithFork(MultiProcessTestCase, DistAutogradTest):
+# dist_autograd_fork tests use double as the default dtype
+torch.set_default_dtype(torch.double)
```
Why float32 is not enough?

xush6528 replied:
I don't know the reason. I triggered auto formatter. Reverting these few lines.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -28,26 +38,44 @@ def register_backend(backend_name, init_backend_handler):
     return backend.value.init_backend_handler(*args, **kwargs)
 
 
+def process_group_construct_rpc_agent_options_handler(
```
This is a very long name, does `process_group_agent_options_handler` work here?

xush6528 replied:
Hey, I would keep it as is. 

It's long. Since it's not a user-facing API, I think it's not a pain for users. 

I am making the naming conforming to a schema here. i.e. <backend_name> + <attribute in the BackendValue type> + "handler"

pietern replied:
Please prefix internal functions with an `_` as a hint to readers that it's internal.

xush6528 replied:
Updated.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -66,32 +66,49 @@ def _init_rpc(
         raise RuntimeError("RPC package does not support Python2.")
 
+    if not isinstance(backend, backend_registry.BackendType):
+        raise RuntimeError("`self_name` must be a string.")
```
wrong error message

xush6528 replied:
Thanks for catching it. Updating.

mrshenli commented on test/test_dist_autograd_fork.py:
```diff
@@ -1,21 +1,23 @@
+torch.set_default_dtype(torch.double)
 
+
+class DistAutogradTestWithFork(
```
Would I be correct if I assume this although this class name does not say process group but it will only run test using process group agent? If I add a new backend, I need to create two new classes: 1) `BackendXYZDistAutogradTest`, and 2) `BackendXYZDistAutogradTestWithFork `? 

If the above is true, shall we merge these two classes into one as you did in the first version, but highlight process group in the class name?

(Doesn't need to be addressed in this PR) And if this is just setting the fixture, I wonder whether we should do this using a decorator and keep `DistAutogradTestWithFork` class generic.

xush6528 replied:
The generic test suite to be re-used by other RPC bakcneds is `dist_autograd_test.DistAutogradTest`

I should rename this to be `ProcessGroupDistAutogradForkTest`.

mrshenli reviewed (commented):
@pritamdamania87 Could you please take a look at the changes in dist autograd tests?

pietern commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -12,6 +12,11 @@ namespace torch {
 namespace rpc {
 
+struct ProcessGroupRpcAgentOptions : public RpcAgentOptions {
+  ProcessGroupRpcAgentOptions() {}
```
Default constructor should go, or defined as `= default`.

xush6528 replied:
Updating.

pietern commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -12,6 +12,11 @@ namespace torch {
 namespace rpc {
 
+struct RpcAgentOptions {
+  RpcAgentOptions() {}
```
Same -- constructor should go.

xush6528 replied:
Resolved.

pietern commented on torch/distributed/rpc/__init__.py:
```diff
@@ -50,34 +50,26 @@ def init_model_parallel(
         )
         store, _, _ = next(rendezvous_iterator)
 
+        # Initialize Autograd.
```
Superfluous comment.

xush6528 replied:
Removing.

pietern reviewed (commented):
There is a ton of test refactoring going on in this PR.

Please update the PR summary stating why the test refactoring is needed. I like the idea of making this stuff composable and having a number of test mixins that let you define what and how to run these tests, but that should have been done as a separate PR. The actual meat of this PR that adds the options struct is only about 1/3rd of the total LOC change.

I'd ask you to split it up, but I know it's going to be a ton more work that would land in sequence anyway, but please take this comment to heart for subsequent changes. PRs should have a focused purpose. That makes them easier to review and easier to comprehend if you're spelunking through Git logs.

Regarding the test refactoring part of this PR: the number of files in `test` that are only relevant for RPC is getting out of hand. There are fixtures, base classes, mixins, and actual test files. This is far beyond the original state of just having `test_xyz.py` files in this directory.

Can you work on moving all of this to a subdirectory like `test/distributed/rpc`?

xush6528 committed:
Update on "Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s"


https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_model_parallel` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: [D18283849](https://our.internmc.facebook.com/intern/diff/D18283849/)

[ghstack-poisoned]

pritamdamania87 commented on test/process_group_dist_autograd_test.py:
```diff
@@ -0,0 +1,15 @@
+import process_group_rpc_agent_test_fixture
+
+
+class ProcessGroupDistAutogradTest(
```
Why do we have test files specific to `ProcessGroup`? This would be hard to maintain as we add more RPC backends. Why can't we just have a single test file and enable/disable tests based on backends as we used to do before?

xush6528 replied:
Answered.

pritamdamania87 commented on test/process_group_rpc_agent_test_fixture.py:
```diff
@@ -0,0 +1,17 @@
+    @property
+    def rpc_agent_options(self):
+        return rpc.backend_registry.construct_rpc_agent_options(
+            self.rpc_backend,
```
Where is `self.rpc_backend` initialized?

xush6528 replied:
In `RpcAgentTestFixture`.

pritamdamania87 reviewed (changes requested):
Have a general question about the structure of the test files here. Can we avoid having ProcessGroup specific test fixtures? I'm wondering why we need such a fixture instead of selectively enabling/disabling individual tests for a particular backend.

xush6528 committed:
Update on "Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s"


https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_model_parallel` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: [D18283849](https://our.internmc.facebook.com/intern/diff/D18283849/)

[ghstack-poisoned]

xush6528 commented:
@pietern 
I have split the test refactoring part of this PR into a new PR, #29747.
Also, moved rpc-related tests into `test/distributed/rpc/`.

xush6528 commented:
@pritamdamania87 

This refactoring is necessary to enable different backends to provide different `RpcAgentOptions`.

xush6528 committed:
Update on "Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s"


https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_model_parallel` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: [D18283849](https://our.internmc.facebook.com/intern/diff/D18283849/)

[ghstack-poisoned]

pritamdamania87 reviewed (changes requested):
@xush6528 I'm assuming we'll be updating this PR based on the offline discussion we had where we can move the options to `TEST_CONFIG`?

xush6528 commented:
@pritamdamania87 

Having a ghexport error..

```
  File "ghstack/submit.py", line 635, in process_old_commit
RuntimeError: Something very strange has happened: a commit for the pull request #29747 occurs twice in your local commit stack.  This is usually because of a botched rebase.  Please take a look at your git log and seek help from your local Git expert.
``

Seems It does not support exporting stack that based on an internal only commit. 

I issued a Phabricator export. Probably it will be create a new PR.

xush6528 commented:
Re-exported to #29972
